janus_server/src/metrics.rs

```diff
@@ -0,0 +1,188 @@
+//! Collection and exporting of application-level metrics for Janus.
+
+use opentelemetry::{
+    metrics::Descriptor,
+    sdk::export::metrics::{Aggregator, AggregatorSelector},
+};
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, net::AddrParseError, sync::Arc};
+
+/// Errors from initializing metrics provider, registry, and exporter.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("bad IP address: {0}")]
+    IpAddress(#[from] AddrParseError),
+    #[error(transparent)]
+    OpenTelemetry(#[from] opentelemetry::metrics::MetricsError),
+    #[cfg(feature = "otlp")]
+    #[error(transparent)]
+    TonicMetadataKey(#[from] tonic::metadata::errors::InvalidMetadataKey),
+    #[cfg(feature = "otlp")]
+    #[error(transparent)]
+    TonicMetadataValue(#[from] tonic::metadata::errors::InvalidMetadataValue),
+    #[error("{0}")]
+    Other(&'static str),
+}
+
+/// Configuration for collection/exporting of application-level metrics.
+#[derive(Clone, Debug, PartialEq, Eq, Default, Serialize, Deserialize)]
+pub struct MetricsConfiguration {
+    /// Configuration for OpenTelemetry metrics, with a choice of exporters.
+    #[serde(default)]
+    pub exporter: Option<MetricsExporterConfiguration>,
+}
+
+/// Selection of an exporter for OpenTelemetry metrics.
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub enum MetricsExporterConfiguration {
+    #[serde(rename = "prometheus")]
```

Comment:
nit: you could put `#[serde(rename_all = "lowercase"]` on the enum to avoid the repreated `#[serde(rename)]` blocks.